core/node/state_keeper/src/seal_criteria/io_criteria.rs

```diff
@@ -0,0 +1,333 @@
+                 commit deadline: {block_commit_deadline_ms}ms",
+                display_timestamp(manager.batch_timestamp())
+            );
+        }
+        Ok(should_seal_timeout)
+    }
+
+    fn should_seal_l2_block(&mut self, manager: &UpdatesManager) -> bool {
+        !manager.l2_block.executed_transactions.is_empty()
+            && millis_since(manager.l2_block.timestamp) > self.l2_block_commit_deadline_ms
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct L2BlockMaxPayloadSizeSealer {
+    max_payload_size: usize,
+}
+
+impl L2BlockMaxPayloadSizeSealer {
+    pub fn new(config: &StateKeeperConfig) -> Self {
+        Self {
+            max_payload_size: config.l2_block_max_payload_size,
+        }
+    }
+
+    pub fn should_seal_l2_block(&mut self, manager: &UpdatesManager) -> bool {
+        manager.l2_block.payload_encoding_size >= self.max_payload_size
+    }
+}
+
+/// Seals L1 batch if pending protocol upgrade is ready to happen or
+/// if the batch is the first one after protocol upgrade.
+#[derive(Debug)]
+pub(crate) struct ProtocolUpgradeSealer {
+    last_checked_at: Option<Instant>,
+    check_interval: Duration,
+}
+
+impl ProtocolUpgradeSealer {
+    pub fn new() -> Self {
+        const DEFAULT_CHECK_INTERVAL: Duration = Duration::from_secs(10);
+
+        Self {
+            last_checked_at: None,
+            check_interval: DEFAULT_CHECK_INTERVAL,
+        }
+    }
+
+    #[cfg(test)]
+    pub fn with_check_interval(mut self, check_interval: Duration) -> Self {
+        self.check_interval = check_interval;
+        self
+    }
+
+    pub async fn should_seal_l1_batch_unconditionally(
+        &mut self,
+        manager: &UpdatesManager,
+        pool: &ConnectionPool<Core>,
+        previous_batch_protocol_version: ProtocolVersionId,
+    ) -> anyhow::Result<bool> {
+        if manager.pending_executed_transactions_len() > 0
+            && manager.protocol_version() != previous_batch_protocol_version
+        {
+            AGGREGATION_METRICS.l1_batch_reason_inc_criterion("first_batch_after_upgrade");
+            return Ok(true);
+        }
+
+        if self
+            .last_checked_at
+            .is_some_and(|last_check_at| last_check_at.elapsed() < self.check_interval)
+        {
+            return Ok(false);
+        }
+
+        let mut conn = pool.connection_tagged("protocol_upgrade_sealer").await?;
+        let current_timestamp = (millis_since_epoch() / 1_000) as u64;
+        let protocol_version = conn
+            .protocol_versions_dal()
+            .protocol_version_id_by_timestamp(current_timestamp)
+            .await
+            .context("Failed loading protocol version")?;
+        let should_seal = protocol_version != manager.protocol_version();
+        if should_seal {
+            AGGREGATION_METRICS.l1_batch_reason_inc_criterion("last_batch_before_upgrade");
+        }
+        self.last_checked_at = Some(Instant::now());
+
+        Ok(should_seal)
```

Comment:
Shouldn't it be checked that the sealed batch is non-empty? Or is it checked implicitly somewhere above?

Reply:
Nice catch, it should be checked